uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,794 @@
+// This file was autogenerated by some hot garbage in the `uniffi` crate.
+// Trust me, you don't want to mess with it!
+
+#ifndef mozilla_dom_{{ ci.namespace()|class_name_webidl }}_Shared
+#define mozilla_dom_{{ ci.namespace()|class_name_webidl }}_Shared
+
+#include <functional>
+
+#include "nsTArray.h"
+#include "prnetdb.h"
+
+#include "mozilla/Casting.h"
+#include "mozilla/CheckedInt.h"
+#include "mozilla/ErrorResult.h"
+#include "mozilla/Utf8.h"
+
+#include "mozilla/dom/BindingDeclarations.h"
+#include "mozilla/dom/Record.h"
+#include "mozilla/dom/{{ ci.namespace()|class_name_webidl }}Binding.h"
+
+extern "C" {
+
+struct RustBuffer {
+  int64_t mLen;
+  uint8_t* mData;
+};
+
+struct RustError {
+  int32_t mCode;
+  char* mMessage;
+};
+
+{% for func in ci.iter_ffi_function_definitions() -%}
+{%- match func.return_type() -%}
+{%- when Some with (type_) %}
+{{ type_|type_ffi }}
+{% when None %}
+void
+{%- endmatch %}
+{{ func.name() }}(
+    {%- for arg in func.arguments() %}
+    {{ arg.type_()|type_ffi }} {{ arg.name() -}}{%- if loop.last -%}{%- else -%},{%- endif -%}
+    {%- endfor %}
+    {%- if func.has_out_err() -%}{%- if func.arguments().len() > 0 %},{% endif %}
+    RustError* out_err
+    {%- endif %}
+);
+
+{% endfor -%}
+
+}  // extern "C"
+
+namespace mozilla {
+namespace dom {
+
+// TODO: Rename this to something less conflict-y, and
+// make sure `lift_cpp` and `lower_cpp` know about it.
+namespace detail {
+
+/// A helper class to read values out of a Rust byte buffer.
+class MOZ_STACK_CLASS Reader final {
+ public:
+  explicit Reader(const RustBuffer& aBuffer) : mBuffer(aBuffer), mOffset(0) {}
+
+  /// Indicates if the offset has reached the end of the buffer.
+  bool HasRemaining() {
+    return static_cast<int64_t>(mOffset.value()) < mBuffer.mLen;
+  }
+
+  /// Helpers to read fixed-width primitive types at the current offset.
+  /// Fixed-width integers are read in big endian order.
```

Comment:
I always find endianness tricky to think about, and I like to document things carefully. With that in mind, for me this comment raises as many questions as it answers. How/why are they read in big-endian order? Is big-endian used for all serialization in `uniffi-rs`? Some more explanation would be helpful.

(It might not hurt to mention that "network order" == "big-endian"... I had to google that to remind myself.)

Reply:
> Is big-endian used for all serialization in uniffi-rs?

It is, and it's probably worth a separate doc for the serialization format as a whole, since it's rather ad-hoc.

Some light background that might only be known to myself: I arbitrarily decided that everything should be big-endian when writing the first version of the serialization code, and I've no particular attachment to that decision. I wonder if it would be better to explicitly commit to using the native machine endianness, since we only ever expect to be serializing between two codebases running on the same machine (in the same process no less).

Reply:
I clarified in the functions themselves that this is because the serialization format uses big endian (I don't think the caller needs to know that one way or the other, any more than it needs to know we serialize floats and doubles by their bit pattern), and I also agree about documenting the format, maybe as part of the [ADR that introduced it](https://github.com/mozilla/uniffi-rs/blob/757a028278f8c8d81754e6a32ea3844012669deb/docs/adr/0002-serialize-complex-datatypes.md), or as an appendix in the tutorial! 👍🏻